ledger/simulation/tracer.go

```diff
@@ -406,6 +412,27 @@ func (tracer *evalTracer) BeforeProgram(cx *logic.EvalContext) {
 		}
 	}
 
+	if cx.RunMode() == logic.ModeApp && tracer.result.ReturnInitialState() {
+		programBytes := cx.GetProgram()
+		programHash := crypto.Hash(programBytes)
+		tracer.result.InitialState.ProgramBytes[string(programBytes)] = struct{}{}
+
+		var txnInitialState *TxnInitialState
+
+		if len(tracer.txnInitialStateStack) == 0 {
+			tracer.result.TxnGroups[0].Txns[groupIndex].InitialState = &TxnInitialState{}
+			txnInitialState = tracer.result.TxnGroups[0].Txns[groupIndex].InitialState
+		} else {
+			lastInitialState := tracer.txnInitialStateStack[len(tracer.txnInitialStateStack)-1]
+			lastInitialState.InnerTxnInitialStates = append(lastInitialState.InnerTxnInitialStates, TxnInitialState{})
+			txnInitialState = &lastInitialState.InnerTxnInitialStates[len(lastInitialState.InnerTxnInitialStates)-1]
+		}
+
+		txnInitialState.ProgramHash = programHash[:]
+
+		tracer.txnInitialStateStack = append(tracer.txnInitialStateStack, txnInitialState)
+	}
+
```

Comment:
How this func might look if it were in a switch 

```
func (tracer *evalTracer) BeforeProgram(cx *logic.EvalContext) {
	groupIndex := cx.GroupIndex()

	switch cx.RunMode() {
	case logic.ModeSig:
		if tracer.result.ReturnTrace() {
			tracer.result.TxnGroups[0].Txns[groupIndex].Trace = &TransactionTrace{}
			traceRef := tracer.result.TxnGroups[0].Txns[groupIndex].Trace
			traceRef.programTraceRef = &traceRef.LogicSigTrace
		}

	case logic.ModeApp:
		if tracer.result.ReturnInitialState() {
			programBytes := cx.GetProgram()
			programHash := crypto.Hash(programBytes)
			tracer.result.InitialState.ProgramBytes[string(programBytes)] = struct{}{}

			var txnInitialState *TxnInitialState

			if len(tracer.txnInitialStateStack) == 0 {
				tracer.result.TxnGroups[0].Txns[groupIndex].InitialState = &TxnInitialState{}
				txnInitialState = tracer.result.TxnGroups[0].Txns[groupIndex].InitialState
			} else {
				lastInitialState := tracer.txnInitialStateStack[len(tracer.txnInitialStateStack)-1]
				lastInitialState.InnerTxnInitialStates = append(lastInitialState.InnerTxnInitialStates, TxnInitialState{})
				txnInitialState = &lastInitialState.InnerTxnInitialStates[len(lastInitialState.InnerTxnInitialStates)-1]
			}

			txnInitialState.ProgramHash = programHash[:]
			tracer.txnInitialStateStack = append(tracer.txnInitialStateStack, txnInitialState)
		}

		if tracer.unnamedResourcePolicy != nil {
			globalSharing := false
			for iter := cx; iter != nil; iter = iter.GetCaller() {
				if iter.ProgramVersion() >= 9 {
					globalSharing = true
					break
				}
			}
			tracer.unnamedResourcePolicy.globalSharing = globalSharing
			tracer.unnamedResourcePolicy.programVersion = cx.ProgramVersion()
			if tracer.unnamedResourcePolicy.initialBoxSurplusReadBudget == nil {
				s := cx.SurplusReadBudget
				tracer.unnamedResourcePolicy.initialBoxSurplusReadBudget = &s
			}
			cx.SetIOBudget(tracer.unnamedResourcePolicy.tracker.maxPossibleBoxIOBudget(cx.Proto.BytesPerBoxReference))
		}
	}
}

Reply:
fair point, I think this should apply to both `BeforeProgram` and `AfterProgram`, as both are `RunMode` dependent.

Reply:
I think let's merge this in first, and I will resolve this in #5659